std/digest/sha3.d

```diff
@@ -0,0 +1,510 @@
+/**
+ * Computes SHA-3 hashes of arbitary data.
+ *
+ * References: NIST FIPS PUB 202
+ *
+ * Source: $(PHOBOSSRC std/digest/sha3.d)
+ */
+module std.digest.sha3;
+
+public import std.digest;
+
+private import core.bitop : rol, bswap;
+
+private immutable ulong[24] K_RC = [
+    0x0000000000000001, 0x0000000000008082, 0x800000000000808a, 0x8000000080008000,
+    0x000000000000808b, 0x0000000080000001, 0x8000000080008081, 0x8000000000008009,
+    0x000000000000008a, 0x0000000000000088, 0x0000000080008009, 0x000000008000000a,
+    0x000000008000808b, 0x800000000000008b, 0x8000000000008089, 0x8000000000008003,
+    0x8000000000008002, 0x8000000000000080, 0x000000000000800a, 0x800000008000000a,
+    0x8000000080008081, 0x8000000000008080, 0x0000000080000001, 0x8000000080008008
+];
+private immutable int[24] K_ROTC = [
+     1,  3,  6, 10, 15, 21, 28, 36, 45, 55,  2, 14,
+    27, 41, 56,  8, 25, 43, 62, 18, 39, 61, 20, 44
+];
+private immutable int[24] K_PI = [
+    10,  7, 11, 17, 18, 3,  5, 16,  8, 21, 24, 4,
+    15, 23, 19, 13, 12, 2, 20, 14, 22,  9,  6, 1
+];
+
+/**
+ * Template API SHA-3/SHAKE implementation using the Keccak[1600] function.
+ * Supports SHA-3-224, SHA-3-256, SHA-3-384, SHA-3-512, SHAKE-128, and SHAKE-256.
+ *
+ * The digestSize parameter is in bits. However, it's easier to use the SHA3_224,
+ * SHA3_256, SHA3_384, SHA3_512, SHAKE128, and SHAKE256 aliases.
+ */
+public struct KECCAK(uint digestSize, bool shake = false)
+{
+    static if (shake)
+        static assert(digestSize == 128 || digestSize == 256,
+            "digest size must be 128 or 256 bits for SHAKE");
+    else
+        static assert(digestSize == 224 || digestSize == 256 ||
+            digestSize == 384 || digestSize == 512,
+            "digest size must be 224, 256, 384, or 512 bits for SHA-3");
+
+    @safe @nogc pure nothrow:
+
+    enum
+    {
+        blockSize = digestSize, /// digest size in bits
+        dgst_sz_bytes = blockSize >> 3, /// digest size in bytes
+        delim = shake ? 0x1f : 0x06, /// delimiter when finishing
+        rate = 200 - (blockSize >> 2), /// sponge rate
```

Comment:
Elsewhere in std.digest `blockSize` has the meaning you have given to `rate` here.

```suggestion
        blockSize = 200 - (digestSize >> 2), /// sponge rate in bytes
        dgst_sz_bytes = digestSize >> 3, /// digest size in bytes
        delim = shake ? 0x1f : 0x06, /// delimiter when finishing
        private alias rate = blockSize;
```

Reply:
Pushed ahead a mix of this suggestion, marking the enum as `private`, and changed the field name `dgst_sz_bytes` to `digestSizeBytes` to better fit the D Style.